### PR TITLE
tests/builder/amazon: Enhance the test coverage for SSM Tunnel related steps

### PR DESCRIPTION
**Pull request by nywilken**
This PR enhances the test coverage to the already existing SSMDriver and StepCreateSSMTunnel test. It adds tests for the Run, Cleanup, and Start/Stop Session methods that drive the SSM tunnel creation. There is also a test to validate that the code does not try to call TerminateSession in the code cleanup when a session was never started :smile: 

Relates #9203 

Amazon with Provisioner Acceptance Test
```
--- PASS: TestShellProvisioner (246.53s)
--- PASS: TestShellProvisioner/testing_amazon-ebs_builder_against_shell_provisioner (246.53s)                                                 
--- PASS: TestAccPowershellProvisioner_Script (268.97s)
--- PASS: TestAccPowershellProvisioner_Script/testing_amazon_ebs_builder_against_powershell_provisioner (268.97s)
```

Amazon Regression Suite
```
⇶  python -m runner --generated=true --configs="ebs-test-1.json,ebs-test-2.json"
[log information redacted]
all tests passed!
```

### CODE REVIEW

File: builder/amazon/common/ssm_mock_funcs.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+func (svc *MockSSMSvc) StartSessionWithContext(ctx aws.Context, input *ssm.StartSessionInput, options ...request.Option) (*ssm.StartSessionOutput, error) {
+	svc.StartSessionCalled = true
+	return MockStartSessionOutput(), svc.StartSessionError
+}
+func (svc *MockSSMSvc) TerminateSession(input *ssm.TerminateSessionInput) (*ssm.TerminateSessionOutput, error) {
+	svc.TerminateSessionCalled = true
+	return new(ssm.TerminateSessionOutput), svc.TerminateSessionError
+}
+
+type MockSSMDriver struct {
```
Comment (SwampDragons) [inline builder/amazon/common/ssm_mock_funcs.go:29]: This is no longer necessary, right?

File: builder/amazon/common/step_create_ssm_tunnel.go [hunk 2]
```diff
@@ -50,62 +50,40 @@ func (s *StepCreateSSMTunnel) Run(ctx context.Context, state multistep.StateBag)
 	}
 	s.instanceId = aws.StringValue(instance.InstanceId)
 
-	log.Printf("Starting PortForwarding session to instance %q on local port %d to remote port %d", s.instanceId, s.LocalPortNumber, s.RemotePortNumber)
```
Comment (SwampDragons) [inline builder/amazon/common/step_create_ssm_tunnel.go:53]: I'm glad you moves this into the driver; it makes sense to make all connection details go in there.

File: builder/amazon/common/ssm_driver.go [hunk 3]
```diff
@@ -6,39 +6,81 @@ import (
...
+	if err != nil {
+		return nil, fmt.Errorf("error encountered in starting session for instance %q: %s", aws.StringValue(input.Target), err)
+	}
+
+	d.session = output
+	d.sessionParams = input
+
+	if d.pluginCmdFunc == nil {
+		d.pluginCmdFunc = d.openTunnelForSession
+	}
```
Comment (azr) [inline builder/amazon/common/ssm_driver.go:68]: Super nice ! Using funcs like this make for super easy testing 🙂

File: builder/amazon/common/ssm_mock_funcs.go (hunk 1 above)
Comment (nywilken) [reply builder/amazon/common/ssm_mock_funcs.go]: Correct, no longer needed. Removed

**Pull request merged**